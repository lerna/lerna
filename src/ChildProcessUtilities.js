import child from "child_process";
import spawn from "cross-spawn";
import objectAssign from "object-assign";
import syncExec from "sync-exec";
import {EventEmitter} from "events";

// Keep track of how many live children we have.
let children = 0;

// maxBuffer value for running exec
const MAX_BUFFER = 500 * 1024;

// This is used to alert listeners when all children have exited.
const emitter = new EventEmitter;

export default class ChildProcessUtilities {
  static exec(command, opts, callback) {
    const mergedOpts = objectAssign({
      maxBuffer: MAX_BUFFER
    }, opts);
    return ChildProcessUtilities.registerChild(
      child.exec(command, mergedOpts, (err, stdout, stderr) => {
        if (err != null) {

          // If the error from `child.exec` is just that the child process
          // emitted too much on stderr, then that stderr output is likely to
          // be useful.
          if (/^stderr maxBuffer exceeded/.test(err.message)) {
            err = `Error: ${err.message}.  Partial output follows:\n\n${stderr}`;
          }

          callback(err || stderr, stdout);
        } else {
          callback(null, stdout);
        }
      })
    );
  }

  static execSync(command, opts) {
    const mergedOpts = objectAssign({
      encoding: "utf8",
      maxBuffer: MAX_BUFFER
    }, opts);
    if (child.execSync) {
      return child.execSync(command, mergedOpts).trim();
    } else {
      return syncExec(command, mergedOpts).stdout.trim();
    }
  }

  static spawn(command, args, opts, callback) {
    let stderr = "";

    const childProcess = ChildProcessUtilities.registerChild(
      spawn(command, args, objectAssign({
        stdio: "inherit"
      }, opts))
        .on("error", () => {})
        .on("exit", (code) => {
          callback(code && (stderr || `Command failed: ${command} ${args.join(" ")}`));
        })
    );

    // By default stderr is inherited from us (just sent to _our_ output).
    // If the caller overrode that to "pipe", then we'll gather that up and
    // call back with it in case of failure.
    if (childProcess.stderr) {
      childProcess.stderr.setEncoding("utf8");
      childProcess.stderr.on("data", (chunk) => stderr += chunk);
    }
  }

  static registerChild(child) {
    children++;
    child.on("exit", () => {
      children--;
      if (children === 0) {
        emitter.emit("empty");
      }
    });
    return child;
  }

  static getChildProcessCount() {
    return children;
  }

  static onAllExited(callback) {
    emitter.on("empty", callback);
  }
}
