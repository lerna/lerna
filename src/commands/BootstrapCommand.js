import FileSystemUtilities from "../FileSystemUtilities";
import NpmUtilities from "../NpmUtilities";
import PackageUtilities from "../PackageUtilities";
import Command from "../Command";
import async from "async";
import find from "lodash.find";
import path from "path";

export default class BootstrapCommand extends Command {
  initialize(callback) {
    this.configFlags = this.repository.bootstrapConfig;
    callback(null, true);
  }

  execute(callback) {
    this.bootstrapPackages((err) => {
      if (err) {
        callback(err);
      } else {
        this.logger.success(`Successfully bootstrapped ${this.filteredPackages.length} packages.`);
        callback(null, true);
      }
    });
  }

  /**
   * Bootstrap packages
   * @param {Function} callback
   */
  bootstrapPackages(callback) {
    this.filteredGraph = PackageUtilities.getPackageGraph(this.filteredPackages);
    this.logger.info(`Bootstrapping ${this.filteredPackages.length} packages`);
    async.series([
      // preinstall bootstrapped packages
      (cb) => this.preinstallPackages(cb),
      // install external dependencies
      (cb) => this.installExternalDependencies(cb),
      // symlink packages and their binaries
      (cb) => this.symlinkPackages(cb),
      // postinstall bootstrapped packages
      (cb) => this.postinstallPackages(cb),
      // prepublish bootstrapped packages
      (cb) => this.prepublishPackages(cb)
    ], callback);
  }


  runScriptInPackages(scriptName, callback) {
    const packages = this.filteredPackages.slice();
    const batches = PackageUtilities.topologicallyBatchPackages(packages, this.logger);

    this.progressBar.init(packages.length);

    const bootstrapBatch = () => {
      const batch = batches.shift();

      async.parallelLimit(batch.map((pkg) => (done) => {
        pkg.runScript(scriptName, (err) => {
          this.progressBar.tick(pkg.name);
          done(err);
        });
      }), this.concurrency, (err) => {
        if (batches.length && !err) {
          bootstrapBatch();
        } else {
          this.progressBar.terminate();
          callback(err);
        }
      });
    };

    // Kick off the first batch.
    bootstrapBatch();
  }

  /**
   * Run the "preinstall" NPM script in all bootstrapped packages
   * @param callback
   */
  preinstallPackages(callback) {
    this.logger.info("Preinstalling packages");
    this.runScriptInPackages("preinstall", callback);
  }

  /**
   * Run the "postinstall" NPM script in all bootstrapped packages
   * @param callback
   */
  postinstallPackages(callback) {
    this.logger.info("Postinstalling packages");
    this.runScriptInPackages("postinstall", callback);
  }

  /**
   * Run the "prepublish" NPM script in all bootstrapped packages
   * @param callback
   */
  prepublishPackages(callback) {
    this.logger.info("Prepublishing packages");
    this.runScriptInPackages("prepublish", callback);
  }

  /**
   * Create a symlink to a dependency's binary in the node_modules/.bin folder
   * @param {String} src
   * @param {String} dest
   * @param {String} name
   * @param {String|Object} bin
   * @param {Function} callback
   */
  createBinaryLink(src, dest, name, bin, callback) {
    const destBinFolder = path.join(dest, ".bin");
    // The `bin` in a package.json may be either a string or an object.
    // Normalize to an object.
    const bins = typeof bin === "string"
      ? { [name]: bin }
      : bin;
    const srcBinFiles = [];
    const destBinFiles = [];
    Object.keys(bins).forEach((name) => {
      srcBinFiles.push(path.join(src, bins[name]));
      destBinFiles.push(path.join(destBinFolder, name));
    });
    // make sure when have a destination folder (node_modules/.bin)
    const actions = [(cb) => FileSystemUtilities.mkdirp(destBinFolder, cb)];
    // symlink each binary
    srcBinFiles.forEach((binFile, idx) => {
      actions.push((cb) => FileSystemUtilities.symlink(binFile, destBinFiles[idx], "exec", cb));
    });
    async.series(actions, callback);
  }

  /**
   * Install external dependencies for all packages
   * @param {Function} callback
   */
  installExternalDependencies(callback) {
    this.logger.info("Installing external dependencies");
    this.progressBar.init(this.filteredPackages.length);
    const actions = [];
    this.filteredPackages.forEach((pkg) => {
      const allDependencies = pkg.allDependencies;
      const externalPackages = Object.keys(allDependencies)
        .filter((dependency) => {
          const match = find(this.packages, (pkg) => {
            return pkg.name === dependency;
          });
          return !(match && pkg.hasMatchingDependency(match));
        })
        .filter((dependency) => !pkg.hasDependencyInstalled(dependency))
        .map((dependency) => dependency + "@" + allDependencies[dependency]);
      if (externalPackages.length) {
        actions.push((cb) => NpmUtilities.installInDir(pkg.location, externalPackages, (err) => {
          this.progressBar.tick(pkg.name);
          cb(err);
        }));
      }
    });
    async.parallelLimit(actions, this.concurrency, (err) => {
      this.progressBar.terminate();
      callback(err);
    });
  }

  /**
   * Symlink all packages to the packages/node_modules directory
   * Symlink package binaries to dependent packages' node_modules/.bin directory
   * @param {Function} callback
   */
  symlinkPackages(callback) {
    this.logger.info("Symlinking packages and binaries");
    this.progressBar.init(this.filteredPackages.length);
    const actions = [];
    this.filteredPackages.forEach((filteredPackage) => {
      // actions to run for this package
      const packageActions = [];
      Object.keys(filteredPackage.allDependencies)
        // filter out external dependencies and incompatible packages
        .filter((dependency) => {
          const match = this.packageGraph.get(dependency);
          return match && filteredPackage.hasMatchingDependency(match.package);
        })
        .forEach((dependency) => {
          // get Package of dependency
          const dependencyPackage = this.packageGraph.get(dependency).package;
          // get path to dependency and its scope
          const { location: dependencyLocation } = dependencyPackage;
          const dependencyPackageJsonLocation = path.join(dependencyLocation, "package.json");
          // ignore dependencies without a package.json file
          if (!FileSystemUtilities.existsSync(dependencyPackageJsonLocation)) {
            this.logger.error(
              `Unable to find package.json for ${dependency} dependency of ${filteredPackage.name},  ` +
              "Skipping..."
            );
          } else {
            // get the destination directory name of the dependency
            const pkgDependencyLocation = path.join(filteredPackage.nodeModulesLocation, dependencyPackage.name);
            // check if dependency is already installed
            if (FileSystemUtilities.existsSync(pkgDependencyLocation)) {
              const isDepSymlink = FileSystemUtilities.isSymlink(pkgDependencyLocation);
              // installed dependency is a symlink pointing to a different location
              if (isDepSymlink !== false && isDepSymlink !== dependencyLocation) {
                this.logger.warning(
                  `Symlink already exists for ${dependency} dependency of ${filteredPackage.name}, ` +
                  "but links to different location. Replacing with updated symlink..."
                );
              // installed dependency is not a symlink
              } else if (isDepSymlink === false) {
                this.logger.warning(
                  `${dependency} is already installed for ${filteredPackage.name}. ` +
                  "Replacing with symlink..."
                );
                // remove installed dependency
                packageActions.push((cb) => FileSystemUtilities.rimraf(pkgDependencyLocation, cb));
              }
            }
            // ensure destination path
            packageActions.push((cb) => FileSystemUtilities.mkdirp(
              pkgDependencyLocation.split(path.sep).slice(0, -1).join(path.sep), cb
            ));
            // create package symlink
            packageActions.push((cb) => FileSystemUtilities.symlink(
              dependencyLocation, pkgDependencyLocation, "junction", cb
            ));
            const dependencyPackageJson = require(dependencyPackageJsonLocation);
            if (dependencyPackageJson.bin) {
              const destFolder = filteredPackage.nodeModulesLocation;
              packageActions.push((cb) => {
                this.createBinaryLink(dependencyLocation, destFolder, dependency, dependencyPackageJson.bin, cb);
              });
            }
          }
        });
      actions.push((cb) => {
        async.series(packageActions, (err) => {
          this.progressBar.tick(filteredPackage.name);
          cb(err);
        });
      });
    });
    async.series(actions, (err) => {
      this.progressBar.terminate();
      callback(err);
    });
  }
}
