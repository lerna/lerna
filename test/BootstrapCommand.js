import pathExists from "path-exists";
import assert from "assert";
import path from "path";

import ChildProcessUtilities from "../src/ChildProcessUtilities";
import BootstrapCommand from "../src/commands/BootstrapCommand";
import exitWithCode from "./_exitWithCode";
import initFixture from "./_initFixture";
import stub from "./_stub";

import assertStubbedCalls from "./_assertStubbedCalls";

const STDIO_OPT = ["ignore", "ignore", "pipe"];

describe("BootstrapCommand", function () {
  this.timeout(5000);

  describe("dependencies between packages in the repo", () => {
    let testDir;

    beforeEach(done => {
      testDir = initFixture("BootstrapCommand/basic", done);
    });

    it("should bootstrap files", done => {
      const bootstrapCommand = new BootstrapCommand([], {});

      bootstrapCommand.runValidations();
      bootstrapCommand.runPreparations();

      assertStubbedCalls([
        [ChildProcessUtilities, "spawn", { nodeCallback: true }, [
          { args: ["npm", ["install", "package-1@^0.0.0"], { cwd: path.join(testDir, "packages/package-4"), stdio: STDIO_OPT }] }
        ]]
      ]);

      bootstrapCommand.runCommand(exitWithCode(0, err => {
        if (err) return done(err);

        try {
          assert.ok(!pathExists.sync(path.join(testDir, "lerna-debug.log")), "lerna-debug.log should not exist");

          assert.ok(pathExists.sync(path.join(testDir, "packages/package-2/node_modules")));
          assert.ok(pathExists.sync(path.join(testDir, "packages/package-3/node_modules")));

          assert.ok(pathExists.sync(path.join(testDir, "packages/package-2/node_modules/package-1")));
          assert.ok(pathExists.sync(path.join(testDir, "packages/package-2/node_modules/package-1/index.js")));
          assert.ok(pathExists.sync(path.join(testDir, "packages/package-2/node_modules/package-1/package.json")));

          // Make sure the `prepublish` script got run (index.js got created).
          assert.equal(require(path.join(testDir, "packages/package-2/node_modules/package-1")), "OK");

          assert.ok(pathExists.sync(path.join(testDir, "packages/package-3/node_modules/package-2")));
          assert.ok(pathExists.sync(path.join(testDir, "packages/package-3/node_modules/package-2/package.json")));
          assert.ok(pathExists.sync(path.join(testDir, "packages/package-3/node_modules/package-2/node_modules/package-1/index.js")));

          // Should not exist because mis-matched version
          assert.ok(!pathExists.sync(path.join(testDir, "packages/package-4/node_modules/package-1")));

          done();
        } catch (err) {
          done(err);
        }
      }));
    });

    it("should not bootstrap an ignored package", done => {
      const bootstrapCommand = new BootstrapCommand([], {
        ignore: "package-5"
      });

      bootstrapCommand.runValidations();
      bootstrapCommand.runPreparations();

      assertStubbedCalls([
        [ChildProcessUtilities, "spawn", { nodeCallback: true }, [
          { args: ["npm", ["install", "package-1@^0.0.0"], { cwd: path.join(testDir, "packages/package-4"), stdio: STDIO_OPT }] }
        ]]
      ]);

      bootstrapCommand.runCommand(exitWithCode(0, err => {
        if (err) return done(err);

        try {
          assert.ok(!pathExists.sync(path.join(testDir, "lerna-debug.log")), "lerna-debug.log should not exist");
          assert.ok(!pathExists.sync(path.join(testDir, "packages/package-5/node_modules/package-1")));
          done();
        } catch (err) {
          done(err);
        }
      }));
    });

    it("should not bootstrap ignored packages", done => {
      const bootstrapCommand = new BootstrapCommand([], {
        ignore: "package-@(3|4)"
      });

      bootstrapCommand.runValidations();
      bootstrapCommand.runPreparations();

      bootstrapCommand.runCommand(exitWithCode(0, err => {
        if (err) return done(err);

        try {
          assert.ok(!pathExists.sync(path.join(testDir, "lerna-debug.log")), "lerna-debug.log should not exist");
          assert.ok(!pathExists.sync(path.join(testDir, "packages/package-3/node_modules/package-2")));
          done();
        } catch (err) {
          done(err);
        }
      }));
    });
  });

  describe("external dependencies that haven't been installed", () => {
    let testDir;

    beforeEach(done => {
      testDir = initFixture("BootstrapCommand/cold", done);
    });

    it("should get installed", (done) => {
      const bootstrapCommand = new BootstrapCommand([], {});

      bootstrapCommand.runValidations();
      bootstrapCommand.runPreparations();

      let installed = false;
      stub(ChildProcessUtilities, "spawn", (command, args, options, callback) => {
        assert.deepEqual(args, ["install", "is-positive@^3.1.0"])
        assert.deepEqual(options, { cwd: path.join(testDir, "packages/package-1"), stdio: STDIO_OPT })
        installed = true;
        callback();
      });

      bootstrapCommand.runCommand(exitWithCode(0, err => {
        if (err) return done(err);

        try {
          assert.ok(!pathExists.sync(path.join(testDir, "lerna-debug.log")), "lerna-debug.log should not exist");
          assert.ok(installed, "The external dependency was installed");

          done();
        } catch (err) {
          done(err);
        }
      }));
    });
  });

  describe("external dependencies that have already been installed", () => {
    let testDir;

    beforeEach(done => {
      testDir = initFixture("BootstrapCommand/warm", done);
    });

    it("should not get re-installed", done => {
      const bootstrapCommand = new BootstrapCommand([], {});

      bootstrapCommand.runValidations();
      bootstrapCommand.runPreparations();

      let installed = false;
      stub(ChildProcessUtilities, "spawn", (command, args, options, callback) => {
        installed = true;
        callback();
      });

      bootstrapCommand.runCommand(exitWithCode(0, err => {
        if (err) return done(err);

        try {
          assert.ok(!pathExists.sync(path.join(testDir, "lerna-debug.log")), "lerna-debug.log should not exist");
          assert.ok(!installed, "The external dependency was not installed");

          done();
        } catch (err) {
          done(err);
        }
      }));
    });
  });
});
