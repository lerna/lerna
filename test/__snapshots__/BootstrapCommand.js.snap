// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BootstrapCommand lifecycle scripts should run preinstall, postinstall and prepublish scripts 1`] = `
Object {
  "packages/package-postinstall": Array [
    "postinstall",
  ],
  "packages/package-preinstall": Array [
    "preinstall",
  ],
  "packages/package-prepublish": Array [
    "prepublish",
  ],
}
`;

exports[`BootstrapCommand with at least one external dependency to install should install all dependencies 1`] = `
Object {
  "packages/package-1": Array [
    "external-1@^1.0.0",
    "external-2@^1.0.0",
  ],
}
`;

exports[`BootstrapCommand with external dependencies that haven't been installed should get installed 1`] = `
Object {
  "packages/package-1": Array [
    "external@^1.0.0",
  ],
  "packages/package-2": Array [
    "external@^2.0.0",
  ],
}
`;

exports[`BootstrapCommand with hoisting should hoist 1`] = `
Object {
  "": Array [
    "foo@^1.0.0",
    "@test/package-1@^0.0.0",
  ],
  "packages/package-3": Array [
    "foo@0.1.12",
  ],
}
`;

exports[`BootstrapCommand with hoisting should hoist 2`] = `
Array [
  "packages/package-1/node_modules/foo",
  "packages/package-2/node_modules/foo",
  "packages/package-4/node_modules/@test/package-1",
]
`;

exports[`BootstrapCommand with hoisting should not hoist when disallowed 1`] = `
Object {
  "": Array [
    "foo@^1.0.0",
  ],
  "packages/package-3": Array [
    "foo@0.1.12",
  ],
  "packages/package-4": Array [
    "@test/package-1@^0.0.0",
  ],
}
`;

exports[`BootstrapCommand with hoisting should not hoist when disallowed 2`] = `
Array [
  "packages/package-1/node_modules/foo",
  "packages/package-2/node_modules/foo",
]
`;

exports[`BootstrapCommand with local package dependencies should bootstrap packages 1`] = `
Object {
  "packages/package-1": Array [
    "foo@^1.0.0",
  ],
  "packages/package-2": Array [
    "foo@^1.0.0",
  ],
  "packages/package-3": Array [
    "foo@0.1.12",
  ],
  "packages/package-4": Array [
    "@test/package-1@^0.0.0",
  ],
}
`;

exports[`BootstrapCommand with local package dependencies should bootstrap packages 2`] = `
Array [
  Object {
    "_src": "packages/package-1",
    "dest": "packages/package-2/node_modules/@test/package-1",
    "type": "junction",
  },
  Object {
    "_src": "packages/package-1",
    "dest": "packages/package-3/node_modules/@test/package-1",
    "type": "junction",
  },
  Object {
    "_src": "packages/package-2",
    "dest": "packages/package-3/node_modules/@test/package-2",
    "type": "junction",
  },
  Object {
    "_src": "packages/package-2/cli.js",
    "dest": "packages/package-3/node_modules/.bin/package-2",
    "type": "exec",
  },
  Object {
    "_src": "packages/package-3",
    "dest": "packages/package-4/node_modules/package-3",
    "type": "junction",
  },
  Object {
    "_src": "packages/package-3/cli1.js",
    "dest": "packages/package-4/node_modules/.bin/package3cli1",
    "type": "exec",
  },
  Object {
    "_src": "packages/package-3/cli2.js",
    "dest": "packages/package-4/node_modules/.bin/package3cli2",
    "type": "exec",
  },
]
`;

exports[`BootstrapCommand with local package dependencies should not bootstrap ignored packages 1`] = `
Object {
  "packages/package-1": Array [
    "foo@^1.0.0",
  ],
  "packages/package-2": Array [
    "foo@^1.0.0",
  ],
}
`;

exports[`BootstrapCommand with local package dependencies should only bootstrap scoped packages 1`] = `
Object {
  "packages/package-3": Array [
    "foo@0.1.12",
  ],
  "packages/package-4": Array [
    "@test/package-1@^0.0.0",
  ],
}
`;

exports[`BootstrapCommand with local package dependencies should only bootstrap scoped packages 2`] = `
Array [
  Object {
    "_src": "packages/package-1",
    "dest": "packages/package-3/node_modules/@test/package-1",
    "type": "junction",
  },
  Object {
    "_src": "packages/package-2",
    "dest": "packages/package-3/node_modules/@test/package-2",
    "type": "junction",
  },
  Object {
    "_src": "packages/package-2/cli.js",
    "dest": "packages/package-3/node_modules/.bin/package-2",
    "type": "exec",
  },
  Object {
    "_src": "packages/package-3",
    "dest": "packages/package-4/node_modules/package-3",
    "type": "junction",
  },
  Object {
    "_src": "packages/package-3/cli1.js",
    "dest": "packages/package-4/node_modules/.bin/package3cli1",
    "type": "exec",
  },
  Object {
    "_src": "packages/package-3/cli2.js",
    "dest": "packages/package-4/node_modules/.bin/package3cli2",
    "type": "exec",
  },
]
`;

exports[`BootstrapCommand with multiple package locations should bootstrap any dependencies excluded by --ignore when --include-filtered-dependencies is true 1`] = `
Object {
  "packages/package-1": Array [
    "foo@^1.0.0",
  ],
  "packages/package-2": Array [
    "foo@^1.0.0",
  ],
}
`;

exports[`BootstrapCommand with multiple package locations should bootstrap any dependencies not included by --scope when --include-filtered-dependencies is true 1`] = `
Object {
  "packages/package-1": Array [
    "foo@^1.0.0",
  ],
  "packages/package-2": Array [
    "foo@^1.0.0",
  ],
}
`;

exports[`BootstrapCommand with multiple package locations should bootstrap packages 1`] = `
Object {
  "package-3": Array [
    "foo@0.1.12",
  ],
  "packages/package-1": Array [
    "foo@^1.0.0",
  ],
  "packages/package-2": Array [
    "foo@^1.0.0",
  ],
  "packages/package-4": Array [
    "@test/package-1@^0.0.0",
  ],
}
`;

exports[`BootstrapCommand with multiple package locations should bootstrap packages 2`] = `
Object {
  "packages/package-1": Array [
    "prepublish",
  ],
}
`;

exports[`BootstrapCommand with multiple package locations should bootstrap packages 3`] = `
Array [
  Object {
    "_src": "packages/package-1",
    "dest": "packages/package-2/node_modules/@test/package-1",
    "type": "junction",
  },
  Object {
    "_src": "package-3",
    "dest": "packages/package-4/node_modules/package-3",
    "type": "junction",
  },
  Object {
    "_src": "package-3/cli1.js",
    "dest": "packages/package-4/node_modules/.bin/package3cli1",
    "type": "exec",
  },
  Object {
    "_src": "package-3/cli2.js",
    "dest": "packages/package-4/node_modules/.bin/package3cli2",
    "type": "exec",
  },
  Object {
    "_src": "packages/package-1",
    "dest": "package-3/node_modules/@test/package-1",
    "type": "junction",
  },
  Object {
    "_src": "packages/package-2",
    "dest": "package-3/node_modules/package-2",
    "type": "junction",
  },
  Object {
    "_src": "packages/package-2/cli.js",
    "dest": "package-3/node_modules/.bin/package-2",
    "type": "exec",
  },
]
`;

exports[`BootstrapCommand with multiple package locations should not bootstrap ignored packages 1`] = `
Object {
  "packages/package-1": Array [
    "foo@^1.0.0",
  ],
  "packages/package-2": Array [
    "foo@^1.0.0",
  ],
}
`;

exports[`BootstrapCommand with registry config should install packages from registry 1`] = `
Object {
  "packages/package-1": Array [
    "foo@^1.0.0",
  ],
}
`;
